spec/eof.md

```diff
@@ -161,11 +159,28 @@ Code executing within an EOF environment will behave differently than legacy cod
 - The instruction `JUMPDEST` is renamed to `NOP` and remains charging 1 gas without any effect.
     - Note: jumpdest-analysis is not performed anymore.
 - EOF contract may not deploy legacy code (it is naturally rejected on the code validation stage)
-- Legacy creation transactions (any tranactions with empty `to`) are invalid in case `data` contains EOF code (starts with `EF00` magic)
 - If instructions `CREATE` and `CREATE2` have EOF code as initcode (starting with `EF00` magic)
     - deployment fails (returns 0 on the stack)
     - caller's nonce is not updated and gas for initcode execution is not consumed
 
+#### Creation transactions
+
+Creation transactions (tranactions with empty `to`), with `data` containing EOF code (starting with `EF00` magic) are interpreted as having an EOF `initcontainer` in the `data` and:
+
+- intrinsic gas cost rules and limits defined in EIP-3860 for legacy creation transaction apply. The entire `initcontainer` is treated as initcode
+- **the initcontainer and all its subcontainers are validated recursively**
+- it is checked if the initcode container has its `len(data_section)` equal to `data_size`, i.e. data section content is exactly as the size declared in the header (see [Data section lifecycle](#data-section-lifecycle))
```

Comment:
This should be relaxed. The goal is to allow existing tooling to append constructor arguments as they do now, without any other change. I.e. somehow it should be allowed data section in the initcontainer to be longer than declared.

Alternative is to consider this constructor arguments "stray data" out of initcontainer, and transaction processing would pass it as calldata to initcontainer execution.

Reply:
This is what the:

> There are alternative ideas to this where "stray data" beyond the end of the declared data_section becomes calldata for the initcode execution frame.

bit in the gh description says. That is your second paragraph:

> Alternative is to consider this constructor arguments "stray data" out of initcontainer, and transaction processing would pass it as calldata to initcontainer execution.

I think that is cleaner than rewriting the data section of the initcontainer to include the stray data.

That being said, the PR diff itself right now takes on the model where the tooling would need to include the constructor args in the "proper" data section. I don't have a strong opinion on whether that is better. However, I think Charles mentioned somewhere on discord that tooling will not have a problem with either version, so we should just keep the one which is the cleanest and most consistent from our (EVM/EOF) standpoint.

Reply:
I think asking the tooling to adapt defeats the entire purpose of using old transaction types for this. If the tooling needs to change,  we could just make this a new type 5 transaction without all the hacks we need to fit EOF into old transaction types (empty `to`, input in data section, having to live with CREATE address scheme)

Reply:
Okay, good point, then lets make

> Alternative is to consider this constructor arguments "stray data" out of initcontainer, and transaction processing would pass it as calldata to initcontainer execution.

the baseline. I'll make adjustments to the PR

Reply:
8c571db7

Reply:
@gumb0 @shemnon When implementing in `evmone`, I noticed this alternative has an ugly consequence in EOF validation. The initcontainer sent to 0x0 with some extra bytes (to become calldata / constructor args) doesn't pass a validation rule we currently have, that there are no bytes in the container beyond what the `data_size` would indicate (there can be less bytes for deployed container, but not more!). This rule is of course very useful and desired for all other cases of EOF validation.

Making an exception from this rule for legacy-eof-creation-txs is not ideal (I am experimenting with this rn). Did you consider the 2nd alternative I posted earlier? This:

> reuse the type 4 InitcodeTransaction zeroth field for fetching the initcontainer in such transactions and keep calldata behaving normally.

So here an InitcodeTransaction sent to a 0x0 address causes the initcontainer to be fetched from `initcodes[0]` and tx' calldata functions normally. WDYT?

Reply:
> The initcontainer sent to 0x0 with some extra bytes (to become calldata / constructor args) doesn't pass a validation rule we currently have, that there are no bytes in the container beyond what the `data_size` would indicate (there can be less bytes for deployed container, but not more!).

Did you mean "sent with empty to" instead of `sent to 0x0` here? (otherwise I don't see what's the significance of address 0x0)

This conflict with a validation rule feels like a technicality to me, because entire `data` field of transaction can be seen as concatenation of EOF container + calldata, where EOF container is valid without stray data.

In evmone implementation I think I see how this can cause some problems, maybe some refactoring needed.

Reply:
> Did you mean "sent with empty to" instead of `sent to 0x0` here? (otherwise I don't see what's the significance of address 0x0)

yeah, brainfart on my part, fixed

> This conflict with a validation rule feels like a technicality to me, because entire data field of transaction can be seen as concatenation of EOF container + calldata, where EOF container is valid without stray data.

But "EOF container + calldata" is what we know when we concatenate bytes, from the standpoint of EOF validating a sequence of bytes in `data` it is indistinguishable from an invalid EOF container (if we include the said rule to apply for all containers).

And what do you think about the "InitcodeTransaction -> empty address" alternative per se?

Reply:
> > Did you mean "sent with empty to" instead of `sent to 0x0` here? (otherwise I don't see what's the significance of address 0x0)
> 
> yeah, brainfart on my part, fixed
> 
> > This conflict with a validation rule feels like a technicality to me, because entire data field of transaction can be seen as concatenation of EOF container + calldata, where EOF container is valid without stray data.
> 
> But "EOF container + calldata" is what we know when we concatenate bytes, from the standpoint of EOF validating a sequence of bytes in `data` it is indistinguishable from an invalid EOF container (if we include the said rule to apply for all containers).
 
The idea I had how so far to address this in implementation:
- Validate only the header first, 
- get all section sizes and calculate full container size,
- Get `tx.data[0:container_size]` and validate it as initcode container, 
- `tx.data[container_size:]` is calldata.

But I don't think we need to change the spec much to reflect these implementation details.

> And what do you think about the "InitcodeTransaction -> empty address" alternative per se?

I like this idea, it's cleaner in many ways, but I think decoupling from TXCREATE is valuable.